hydra-node/test/Hydra/HeadLogicSpec.hs

```diff
@@ -543,19 +547,22 @@ getConfirmedSnapshot = \case
     Nothing
 
 -- | Asserts that the update function will update the state (return a NewState) for this Event
-assertUpdateState :: (HasCallStack, IsChainState tx) => Environment -> Ledger tx -> Event tx -> HeadState tx -> IO (HeadState tx)
-assertUpdateState env ledger event st =
-  assertNewState $ update env ledger st event
+assertUpdateState :: (MonadState (HeadState tx) m, HasCallStack, IsChainState tx) => Environment -> Ledger tx -> Event tx -> m (HeadState tx)
+assertUpdateState env ledger event = do
+  st <- get
+  st' <- assertNewState $ update env ledger st event
+  put st'
+  pure st'
 
 assertNewState ::
-  (HasCallStack, IsChainState tx) =>
+  (HasCallStack, IsChainState tx, Applicative m) =>
   Outcome tx ->
-  IO (HeadState tx)
+  m (HeadState tx)
 assertNewState = \case
   NewState st _ -> pure st
-  OnlyEffects effects -> failure $ "Unexpected 'OnlyEffects' outcome: " <> show effects
-  Error e -> failure $ "Unexpected 'Error' outcome: " <> show e
-  Wait r -> failure $ "Unexpected 'Wait' outcome with reason: " <> show r
+  OnlyEffects effects -> Hydra.Prelude.error $ "Unexpected 'OnlyEffects' outcome: " <> show effects
```

Comment:
Could we also use `failure` here? I guess you need to propagate `MonadThrow` then which is a bit annoying.

Reply:
Yes, and that's a PITA because we use io-classes' `MonadThrow` and it does not have `StateT` instance, and I couldn't care to write my own by hand :(